File: java/armstrong-numbers/src/main/java/ArmstrongNumbers.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.IntStream;

class ArmstrongNumbers {

  boolean isArmstrongNumber(int numberToCheck) {
    List<Integer> digits = digits(numberToCheck);
    int numDigits = digits.size();
    int armstrongSum = IntStream.range(0, numDigits).map(i -> pow(digits.get(i), numDigits)).sum();
    return numberToCheck == armstrongSum;
  }

  List<Integer> digits(int number) {
    ArrayList<Integer> digits = new ArrayList<Integer>();
    while (number > 0) {
      digits.add(number % 10);
      number /= 10;
    }
    Collections.reverse(digits);
    return digits;
  }

  int pow(int base, int exponent) {
    return IntStream.range(0, exponent).map(_i -> base).reduce((a, b) -> a * b).getAsInt();
  }
}

File: c/armstrong-numbers/src/armstrong_numbers.c
#include "armstrong_numbers.h"
#include <math.h>

#define BASE 10

static int pow_int(const int base, const int power) {
    int result = 1;
    for (int i = 0; i < power; i++) {
        result *= base;
    }
    return result;
}

int is_armstrong_number(const int candidate) {
    int candidate_copy = candidate;
    int sum = 0;
    const int num_digits = log10(candidate) + 1;
    while (candidate_copy > 0) {
        const int digit = candidate_copy % BASE;
        sum += pow_int(digit, num_digits);
        candidate_copy /= BASE;
    }
    return candidate == sum;
}
